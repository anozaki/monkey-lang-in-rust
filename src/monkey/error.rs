use std::rc::Rc;
use thiserror::Error;
use crate::monkey::token::Token;

#[derive(Error, Debug, PartialEq, Clone)]
pub enum Error {
    #[error("Generic {0}")]
    TodoError(String),

    #[error("Invalid token found at {0:?}")]
    InvalidTokenError(Rc<Token>),

    #[error("Invalid escape char {0:?}")]
    InvalidCharError(char),
}