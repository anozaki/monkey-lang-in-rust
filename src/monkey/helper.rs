pub fn is_identifier(ch: char) -> bool {
    ch.is_ascii_alphabetic() || ch == '_'
}

pub fn is_digit(ch: char) -> bool {
    ch.is_ascii_digit()
}

pub fn is_whitespace(ch: char) -> bool {
    ch.is_whitespace()
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_is_identifier() {
        assert_eq!(is_identifier('a'), true);
        assert_eq!(is_identifier('_'), true);
        assert_eq!(is_identifier('あ'), false);
        assert_eq!(is_identifier('1'), false);
    }

    #[test]
    fn test_is_digit() {
        assert_eq!(is_digit('1'), true);
        assert_eq!(is_digit('2'), true);
        assert_eq!(is_digit('3'), true);
        assert_eq!(is_digit('4'), true);
        assert_eq!(is_digit('_'), false);
        assert_eq!(is_digit('%'), false);
        assert_eq!(is_digit('１'), false); // non ascii 1
    }

    #[test]
    fn test_is_whitespace() {
        assert_eq!(is_whitespace(' '), true);
        assert_eq!(is_whitespace('\n'), true);
        assert_eq!(is_whitespace('\t'), true);
        assert_eq!(is_whitespace('　'), true); // non-ascii space
    }
}