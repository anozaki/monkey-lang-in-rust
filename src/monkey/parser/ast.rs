use std::collections::HashMap;
use std::fmt::{Debug, Display, Formatter};
use std::rc::Rc;

use crate::monkey::token::Token;

#[derive(Debug, Clone, PartialEq)]
pub enum Operator {
    Not,
    Neg,
    Add,
    Sub,
    Mul,
    Div,
    Greater,
    Less,
    Equal,
    NotEqual,
    Call,
    Index,
}

#[derive(Debug, Default, Clone, PartialEq)]
pub struct Program {
    pub statements: Vec<StatementNode>,
}


#[derive(Debug, Clone, PartialEq)]
pub struct Identifier(pub String);

#[derive(Debug, Clone, PartialEq)]
pub enum StatementNode {
    Let(Identifier, Box<ExpressionNode>),
    Return(Box<ExpressionNode>),
    If {
        condition: Box<ExpressionNode>,
        consequence: Box<Program>,
        alternative: Option<Box<Program>>,
    },
    Expression {
        expression: Box<ExpressionNode>
    },
}

#[derive(Debug, Clone, PartialEq)]
pub enum ExpressionNode {
    Identifier(Identifier),
    Bool(bool),
    Int(isize),
    String(String),
    Call {
        function: Box<ExpressionNode>,
        params: Vec<Box<ExpressionNode>>,
    },
    Function {
        params: Vec<Identifier>,
        body: Box<Program>,
    },
    Infix {
        operator: Operator,
        left: Box<ExpressionNode>,
        right: Box<ExpressionNode>,
    },
    Prefix {
        operator: Operator,
        expression: Box<ExpressionNode>,
    },
    ArrayLiteral {
        params: Vec<Box<ExpressionNode>>,
    },
    Index {
        left: Box<ExpressionNode>,
        index: Box<ExpressionNode>,
    },
    HashLiteral {
        params: Vec<(Box<ExpressionNode>, Box<ExpressionNode>)>,
    },
}

pub struct Node {
    token: Rc<Token>,
    program: Program,
}
