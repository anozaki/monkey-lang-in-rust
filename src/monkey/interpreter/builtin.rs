use std::fmt::format;
use crate::monkey::interpreter::{NULL, Object};
use crate::monkey::Result;

pub fn str_len(objs: Vec<Object>) -> Result<Object> {
    let Some(obj) = objs.first() else {
        return Ok(Object::Error(format!("len(): Invalid argument")))
    };

    Ok(match obj {
        Object::String(str) => Object::Int(str.len() as isize),
        Object::Array(arr) => Object::Int(arr.len() as isize),
        _ => Object::Error(format!("len(): Invalid argument: {:?}", obj)),
    })
}

pub fn first(objs: Vec<Object>) -> Result<Object> {
    let Some(obj) = objs.first() else {
        return Ok(Object::Error(format!("first(): Invalid argument")))
    };

    Ok(match obj {
        Object::Array(arr) => match arr.first() {
            None => NULL,
            Some(val) => val.clone()
        },
        _ => Object::Error(format!("first(): Invalid argument: {:?}", obj)),
    })
}

pub fn last(objs: Vec<Object>) -> Result<Object> {
    let Some(obj) = objs.first() else {
        return Ok(Object::Error(format!("last(): Invalid argument")))
    };

    Ok(match obj {
        Object::Array(arr) => match arr.last() {
            None => NULL,
            Some(val) => val.clone()
        },
        _ => Object::Error(format!("last(): Invalid argument: {:?}", obj)),
    })
}

pub fn push(objs: Vec<Object>) -> Result<Object> {
    if objs.len() != 2 {
        return Ok(Object::Error(format!("push(): Invalid argument")))
    }

    let target = &objs[0];
    let obj = &objs[1];

    Ok(match target {
        Object::Array(arr) => {
            let mut result = arr.clone();
            result.push(obj.clone());

            Object::Array(result)
        },
        _ => Object::Error(format!("push(): Invalid argument: {:?}", obj)),
    })
}

pub fn rest(objs: Vec<Object>) -> Result<Object> {
    let Some(obj) = objs.first() else {
        return Ok(Object::Error(format!("rest(): Invalid argument")))
    };

    Ok(match obj {
        Object::Array(arr) => {
            if let Some((last, elements)) = arr.split_last() {
                Object::Array(Vec::from(elements))
            } else {
                Object::Array(vec![])
            }
        },
        _ => Object::Error(format!("rest(): Invalid argument: {:?}", obj)),
    })
}

pub fn put(objs: Vec<Object>) -> Result<Object> {
    if objs.len() != 3 {
        return Ok(Object::Error(format!("put(): Invalid argument")))
    }

    let target = &objs[0];
    let key = &objs[1];
    let val = &objs[2];

    Ok(match target {
        Object::Hash(map) => {
            let mut result = map.clone();
            result.insert(key.clone(), val.clone());

            Object::Hash(result)
        },
        _ => Object::Error(format!("put(): Invalid argument")),
    })
}
