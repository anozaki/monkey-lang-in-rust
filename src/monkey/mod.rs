use crate::monkey::error::Error;

pub mod error;

pub mod lexer;
pub mod helper;
pub mod parser;
pub mod token;


pub mod interpreter;

pub type Result<T> = core::result::Result<T, Error>;
