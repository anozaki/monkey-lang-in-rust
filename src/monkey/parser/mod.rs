use std::rc::Rc;

use crate::monkey::lexer::Lexer;
use crate::monkey::parser::ast::{Program, StatementNode};
use crate::monkey::Result;
use crate::monkey::token::{Token, TokenType};
use crate::monkey::token::TokenType::Semicolon;

pub mod parser;
pub mod expression;
pub mod ast;


#[macro_export]
macro_rules! try_next {
    ($self: ident, $token: expr) => {
        if $self.current().token_type != $token {
            return Err(Error::InvalidTokenError(Rc::clone($self.current())));
        }
        $self.next();
    }
}

fn next_token(lexer: &dyn Lexer) -> Token {
    match lexer.token() {
        Ok(token) => token,
        Err(message) => panic!("Unable to parse: {message}")
    }
}

pub struct Parser<'a> {
    lexer: &'a dyn Lexer,

    token_current: Rc<Token>,
    token_peek: Rc<Token>,
}

impl <'a> Parser <'a> {
    pub fn new(lexer: &'a dyn Lexer) -> Self {
        let token_current = Rc::new(next_token(lexer));
        let token_peek = Rc::new(next_token(lexer));
        Parser {
            lexer,
            token_current,
            token_peek,
        }
    }

    pub fn current(&self) -> &Rc<Token> {
        return &self.token_current;
    }

    pub fn peek(&self) -> &Rc<Token> {
        return &self.token_peek;
    }

    pub fn next(&mut self) {
        self.token_current = Rc::clone(&self.token_peek);
        self.token_peek = Rc::new(next_token(self.lexer));
    }

    pub fn skip_semicolon(&mut self) {
        while self.current().token_type == Semicolon {
            self.next()
        }
    }

    pub fn parse_program(&mut self) -> Result<Box<Program>> {
        let mut program: Box<Program> = Box::new(Program::default());

        while self.current().token_type != TokenType::EndOfFile && self.current().token_type != TokenType::RightBrace {
            program.statements.push(self.parse_statement()?);
            self.skip_semicolon();
        }

        Ok(program)
    }

    pub fn parse_statement(&mut self) -> Result<StatementNode> {
        Ok(match self.current().token_type {
            TokenType::Let => self.parse_let()?,
            TokenType::Return => self.parse_return()?,
            TokenType::If => self.parse_if()?,
            _ => self.parse_expression()?,
        })
    }
}

#[cfg(test)]
mod test {
    use std::fs;
    use crate::monkey::lexer::MonkeyLexer;

    use super::*;

    pub fn evaluate(source: &str) -> String {
        let contents = fs::read_to_string(source).unwrap();
        let lexer = MonkeyLexer::new(&contents);

        let mut parser = Parser::new(&lexer);
        let program = parser.parse_program().unwrap();

        let mut output = String::new();
        for statement in &program.statements {
            output += &format!("{:?}\n", statement);
        }

        output
    }

    macro_rules! evaluate {
        ($name:ident, $file:literal) => {
            #[test]
            fn $name() {
                let output = evaluate(&format!("monkey/{}", $file));
                insta::assert_snapshot!(output)
            }
        }
    }

    evaluate!(test_parse_let, "test_parser_let.mky");
    evaluate!(test_parse_return, "test_parser_return.mky");
    evaluate!(test_parser_infix, "test_parser_infix.mky");
    evaluate!(test_parser_expression, "test_parser_expression.mky");
    evaluate!(test_parser_expression_prec, "test_parser_expression_prec.mky");
    evaluate!(test_parser_if, "test_parser_if.mky");
    evaluate!(test_parser_function_literals, "test_parser_function_literals.mky");
    evaluate!(test_parser_closure, "test_parser_closure.mky");
    evaluate!(test_parser_string, "test_parser_string.mky");
    evaluate!(test_parser_index, "test_parser_index.mky");
    evaluate!(test_parser_hash, "test_parser_hash.mky");
}
