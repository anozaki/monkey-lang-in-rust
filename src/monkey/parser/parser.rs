use std::rc::Rc;

use crate::try_next;
use crate::monkey::error::Error;
use crate::monkey::parser::ast::{Identifier, StatementNode};
use crate::monkey::parser::expression::OrderOfOps;
use crate::monkey::parser::Parser;
use crate::monkey::Result;
use crate::monkey::token::TokenType;

impl <'a> Parser<'a> {
    pub fn parse_let(&mut self) -> Result<StatementNode> {
        let let_token = Rc::clone(&self.current());
        self.next();

        let identifier = Identifier(self.current().literal.clone());
        self.next();

        try_next!(self, TokenType::Assign);

        let expression = self.parse_expression_node(OrderOfOps::Lowest)?;

        self.next();

        Ok(
            StatementNode::Let(
                identifier, expression,
            )
        )
    }

    pub fn parse_if(&mut self) -> Result<StatementNode> {
        let token = Rc::clone(&self.token_current);

        self.next();

        try_next!(self, TokenType::LeftParen);

        let expression = self.parse_expression_node(OrderOfOps::Lowest)?;

        self.next();

        try_next!(self, TokenType::RightParen);

        try_next!(self, TokenType::LeftBrace);

        let consequence = self.parse_program()?;

        try_next!(self, TokenType::RightBrace);

        let alternative = if self.current().token_type == TokenType::Else {
            self.next();

            try_next!(self, TokenType::LeftBrace);

            match self.parse_program() {
                Ok(result) => {
                    self.next();
                    Some(result)
                }
                Err(err) => return Err(err),
            }
        } else {
            None
        };

        Ok(
            StatementNode::If {
                condition: expression,
                consequence,
                alternative,
            }
        )
    }

    pub fn parse_return(&mut self) -> Result<StatementNode> {
        self.next();

        let expression = self.parse_expression_node(OrderOfOps::Lowest)?;

        self.next();

        Ok(StatementNode::Return(expression))
    }
}

