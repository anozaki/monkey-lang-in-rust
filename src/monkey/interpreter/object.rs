use std::cell::RefCell;
use std::collections::HashMap;
use std::fmt::{Debug, Display, Formatter};
use std::hash::{Hash, Hasher};
use std::rc::Rc;

use crate::monkey::interpreter::Environment;
use crate::monkey::parser::ast::{Identifier, Program};
use crate::monkey::Result;

pub type BuiltInFn = fn(Vec<Object>) -> Result<Object>;

#[derive(Clone, PartialEq)]
pub enum Object {
    Null,
    Int(isize),
    Bool(bool),
    Return(Box<Object>),
    Function {
        ident: Vec<Identifier>,
        program: Box<Program>,
        env: Rc<RefCell<Environment>>,
    },
    BuiltIn {
        name: String,
        params: usize,
        program: BuiltInFn,
    },
    String(String),
    Error(String),
    Array(Vec<Object>),
    Hash(HashMap<Object, Object>),
}

// FIXME: is there any easier way to do this?
impl Debug for Object {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Object::Null => write!(f, "Null"),
            Object::Int(val) => write!(f, "Int({:?})", val),
            Object::Bool(val) => write!(f, "Bool({:?})", val),
            Object::Return(val) => write!(f, "Return({:?})", val),

            // env could contain self - to prevent inf recursion we omit printing env.
            Object::Function { ident, program, env } => write!(f, "Function {{ ident: {:?}, program: {:?} }}", ident, program),

            // Omit program since we use this for test and we want predictable output.
            Object::BuiltIn { name, params, program } => write!(f, "BuiltIn {{ name: {:?}, params: {:?} }}", name, params),

            Object::String(val) => write!(f, "String({:?})", val),
            Object::Error(val) => write!(f, "Error({:?})", val),
            Object::Array(val) => write!(f, "Array({:?})", val),
            Object::Hash(result) => write!(f, "Hash({:?})", result),
        }
    }
}

impl Display for Object {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Object::Null => write!(f, "Null"),
            Object::Int(val) => write!(f, "{}", val),
            Object::Bool(val) => write!(f, "{}", val),
            Object::String(val) => write!(f, "\"{}\"", val),
            Object::Error(val) => write!(f, "Error: {}", val),
            Object::Array(val) => {
                write!(f, "[ ");
                let mut first = true;
                for v in val {
                    if !first {
                        write!(f, ", ");
                    }
                    first = false;
                    write!(f, "{}", v);
                }
                write!(f, " ]")
            }
            Object::Hash(val) => {
                write!(f, "{{ ");
                let mut first = true;
                for (key, val) in val {
                    if !first {
                        write!(f, ", ");
                    }
                    first = false;
                    write!(f, "{}: {}", key, val);
                }
                write!(f, " }}")
            }
            val @ _ => Debug::fmt(self, f)
        }
    }
}

impl Eq for Object {}

impl Hash for Object {
    fn hash<H: Hasher>(&self, state: &mut H) {
        match self {
            Object::Int(val) => val.hash(state),
            Object::Bool(val) => val.hash(state),
            Object::String(val) => val.hash(state),

            _ => todo!("Not supported"),
        }
    }
}

pub const TRUE: Object = Object::Bool(true);
pub const FALSE: Object = Object::Bool(false);

pub const NULL: Object = Object::Null;
