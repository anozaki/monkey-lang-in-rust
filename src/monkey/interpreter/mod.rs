use std::cell::RefCell;
use std::collections::HashMap;
use std::fmt::{Debug, Formatter};
use std::hash::{Hash, Hasher};
use std::ptr::hash;
use std::rc::Rc;
use crate::monkey::interpreter::builtin::{first, last, push, put, rest, str_len};
use crate::monkey::interpreter::object::{FALSE, NULL, Object, TRUE};
use crate::monkey::parser::ast::{ExpressionNode, Identifier, Operator, Program, StatementNode};
use crate::monkey::Result;

pub mod builtin;
mod object;


#[derive(Debug, Clone, PartialEq)]
pub struct Environment {
    store: HashMap<String, Object>,
}

impl Environment {
    pub fn new() -> Self {
        let mut store: HashMap<String, Object> = HashMap::new();
        store.insert("len".to_string(), Object::BuiltIn {
            name: "len".to_string(),
            params: 1,
            program: str_len,
        });
        store.insert("first".to_string(), Object::BuiltIn {
            name: "first".to_string(),
            params: 1,
            program: first,
        });
        store.insert("last".to_string(), Object::BuiltIn {
            name: "last".to_string(),
            params: 1,
            program: last,
        });
        store.insert("push".to_string(), Object::BuiltIn {
            name: "push".to_string(),
            params: 2,
            program: push,
        });
        store.insert("rest".to_string(), Object::BuiltIn {
            name: "rest".to_string(),
            params: 1,
            program: rest,
        });
        store.insert("put". to_string(), Object::BuiltIn {
            name: "put".to_string(),
            params: 3,
            program: put,
        });

        Environment {
            store,
        }
    }
    pub fn store(&mut self, name: &str, value: &Object) {
        self.store.insert(name.to_string(), value.clone());
    }

    pub fn get(&self, name: &str) -> Option<Object> {
        match self.store.get(name) {
            None => None,
            Some(val) => Some(val.clone())
        }
    }
}

pub struct Evaluate {}

impl Evaluate {
    pub fn new() -> Self {
        Evaluate {}
    }

    pub fn evaluate(&self, program: &Program) -> Result<Object> {
        let env = Rc::new(RefCell::new(Environment::new()));
        self.evaluate_program(program, &env)
    }

    pub fn evaluate_program(&self, program: &Program, env: &Rc<RefCell<Environment>>) -> Result<Object> {
        let result = self.evaluate_block(program, env)?;
        Ok(match result {
            Object::Return(val) => *val,
            val @ _ => val,
        })
    }

    pub fn evaluate_block(&self, program: &Program, env: &Rc<RefCell<Environment>>) -> Result<Object> {
        let mut result = NULL;

        for statement in &program.statements {
            result = match statement {
                StatementNode::Let(ident, expr) => self.eval_let_statement(ident, expr, env)?,
                StatementNode::If { condition, consequence, alternative } => self.eval_if_statement(condition, consequence, alternative, env)?,
                StatementNode::Return(expression) => Object::Return(Box::new(self.expression(expression, env)?)),
                StatementNode::Expression { expression } => self.expression(expression, env)?,
            };

            if let Object::Return(_) = result {
                break;
            }
        };

        Ok(result)
    }

    pub fn eval_let_statement(&self, identifier: &Identifier, expression: &ExpressionNode, env: &Rc<RefCell<Environment>>) -> Result<Object> {
        let result = self.expression(expression, env)?;
        if matches!(result, Object::Error(_)) {
            return Ok(result);
        }

        env.borrow_mut().store(&identifier.0, &result);

        Ok(result)
    }

    pub fn eval_if_statement(&self, condition: &ExpressionNode, consequence: &Program, alternative: &Option<Box<Program>>, env: &Rc<RefCell<Environment>>) -> Result<Object> {
        let result = self.expression(condition, env)?;

        Ok(
            if self.is_truthy(&result) {
                self.evaluate_block(&consequence, env)?
            } else if let Some(program) = alternative {
                self.evaluate_block(&program, env)?
            } else {
                NULL
            }
        )
    }

    pub fn is_truthy(&self, condition: &Object) -> bool {
        match condition {
            Object::Null => false,
            Object::Int(val) if *val == 0 => false,
            Object::Bool(val) => *val,
            _ => true,
        }
    }

    pub fn expression(&self, expression: &ExpressionNode, env: &Rc<RefCell<Environment>>) -> Result<Object> {
        Ok(match expression {
            ExpressionNode::Int(val) => Object::Int(*val),
            ExpressionNode::Bool(val) => match val {
                true => TRUE,
                false => FALSE,
            }
            ExpressionNode::Identifier(ident) => match env.borrow().get(&ident.0) {
                Some(val) => val,
                None => Object::Error(format!("identifier not found: {:?}", ident))
            },
            ExpressionNode::String(str) => Object::String(str.clone()),
            ExpressionNode::Prefix { operator, expression } => self.eval_prefix(operator, expression, env)?,
            ExpressionNode::Infix { operator, left, right } => {
                let left = self.expression(left, env)?;
                let right = self.expression(right, env)?;
                self.eval_infix(operator, &left, &right)?
            }
            ExpressionNode::Function { params, body } => Object::Function {
                ident: params.clone(),
                program: body.clone(),
                env: Rc::clone(env),
            },
            ExpressionNode::Call { function, params } => {
                let func = self.expression(function, env)?;

                if let Object::Function { ident, program, env: cap } = func {
                    let mut new_env = Environment::new();
                    for (name, obj) in &env.borrow().store {
                        new_env.store(name, obj);
                    }

                    for (_, (ident, item)) in ident.iter().zip(params).enumerate() {
                        let result = self.expression(item, env)?;
                        new_env.store(&ident.0, &result)
                    }

                    self.evaluate_program(&program, &Rc::new(RefCell::new(new_env)))?
                } else if let Object::BuiltIn { name, params: params_size, program } = func {
                    let mut func_params: Vec<Object> = Vec::new();
                    for item in params {
                        func_params.push(self.expression(item, env)?)
                    }

                    if func_params.len() != params_size {
                        return Ok(Object::Error(format!("{}(): Invalid number of argument - expected {} got {}", name, params_size, func_params.len())));
                    }

                    program(func_params)?
                } else if let Object::Error(message) = func {
                    panic!("{}", message)
                } else {
                    panic!()
                }
            }
            ExpressionNode::ArrayLiteral { params } => {
                let mut func_params: Vec<Object> = Vec::new();
                for item in params {
                    func_params.push(self.expression(item, env)?)
                }

                Object::Array(func_params)
            }
            ExpressionNode::Index { left, index } => {
                let left_obj = self.expression(left, env)?;
                match left_obj {
                    Object::Array(vec) => {
                        let index_obj = self.expression(index, env)?;
                        let Object::Int(offset) = &index_obj else {
                            return Ok(Object::Error(format!("Invalid index value: {:?}", index_obj)));
                        };

                        if offset.is_negative() || vec.len() < *offset as usize {
                            return Ok(NULL);
                        }

                        // To keep it simple we will clone...
                        vec[*offset as usize].clone()
                    }
                    Object::Hash(map) => {
                        let index_obj = self.expression(index, env)?;
                        map[&index_obj].clone()

                    },
                    _ => return Ok(Object::Error(format!("Can not index object type: {:?}", left_obj))),
                }

            }
            ExpressionNode::HashLiteral { params} => {
                let mut map: HashMap<Object, Object> = HashMap::new();

                for (key, val) in params {
                    let key_result = self.expression(key, env)?;
                    let val_result = self.expression(val, env)?;

                    map.insert(key_result, val_result);
                }

                Object::Hash(map)
            }
            _ => todo!()
        })
    }

    fn eval_prefix(&self, operator: &Operator, expression: &Box<ExpressionNode>, env: &Rc<RefCell<Environment>>) -> Result<Object> {
        Ok(match operator {
            Operator::Not => match self.expression(expression, env) {
                Ok(result) => match result {
                    Object::Bool(true) => FALSE,
                    Object::Bool(false) => TRUE,
                    Object::Int(0) => TRUE,
                    _ => FALSE,
                },
                Err(_) => panic!()
            },
            Operator::Neg => match self.expression(expression, env) {
                Ok(result) => match result {
                    Object::Int(val) => Object::Int(-val),
                    _ => Object::Error(format!("Unexpected value type: {:?}", result)),
                },
                Err(_) => panic!()
            },
            _ => Object::Error(format!("Unknown operator: {:?}", operator))
        })
    }

    fn eval_infix(&self, operator: &Operator, left: &Object, right: &Object) -> Result<Object> {
        Ok(match (left, right) {
            (Object::Int(left_val), Object::Int(right_val)) => match operator {
                Operator::Add => Object::Int(left_val + right_val),
                Operator::Sub => Object::Int(left_val - right_val),
                Operator::Mul => Object::Int(left_val * right_val),
                Operator::Div => Object::Int(left_val / right_val),
                Operator::Greater => Object::Bool(left_val > right_val),
                Operator::Less => Object::Bool(left_val < right_val),
                Operator::Equal => Object::Bool(left_val == right_val),
                Operator::NotEqual => Object::Bool(left_val != right_val),
                _ => Object::Error(format!("Unexpected Operator: {:?}", operator))
            },
            (Object::String(left_var), Object::String(right_var)) => match operator {
                Operator::Add => Object::String(format!("{}{}", left_var, right_var)),
                _ => Object::Error(format!("Unexpected Operator: {:?}", operator))
            }
            (Object::String(left_var), Object::Int(right_var)) => match operator {
                Operator::Add => Object::String(format!("{}{}", left_var, right_var)),
                _ => Object::Error(format!("Unexpected Operator: {:?}", operator))
            }
            _ => Object::Error(format!("Expected number, got {:?} and {:?}", left, right))
        })
    }
}

#[cfg(test)]
mod test {
    use crate::monkey::lexer::MonkeyLexer;
    use crate::monkey::parser::ast::Operator;
    use crate::monkey::parser::Parser;
    use super::*;

    macro_rules! test_expression {
        ($out: ident, $command: expr) => {
            {
                let lexer = MonkeyLexer::new($command);
                let mut parser = Parser::new(&lexer);

                let program = parser.parse_program().unwrap();

                let mut eval = Evaluate::new();
                let out = eval.evaluate(&program);

                $out += &format!("{:?}\n", out)
            }
        }
    }

    #[test]
    fn text_eval_expression() {
        let mut result = String::new();

        test_expression!(result, "10");
        test_expression!(result, "152");
        test_expression!(result, "-5");
        test_expression!(result, "-1523");

        test_expression!(result, "true");
        test_expression!(result, "false");

        test_expression!(result, "!true");
        test_expression!(result, "!false");
        test_expression!(result, "!5");
        test_expression!(result, "!0");
        test_expression!(result, "!-5");
        test_expression!(result, "!-0");
        test_expression!(result, "!!5");
        test_expression!(result, "!!true");
        test_expression!(result, "!!false");

        test_expression!(result, "5 + 5 + 5 + 5 - 10");
        test_expression!(result, "2 * 2 * 2 * 2 * 2");
        test_expression!(result, "-50 + 100 + -50");
        test_expression!(result, "5 * 2 + 10");
        test_expression!(result, "5 + 2 * 10");
        test_expression!(result, "20 + 2 * -10");
        test_expression!(result, "50 / 2 * 2 + 10");
        test_expression!(result, "2 * (5 + 10)");
        test_expression!(result, "3 * 3 * 3 + 10");
        test_expression!(result, "3 * (3 * 3) + 10");
        test_expression!(result, "(5 + 10 * 2 + 15 / 3) * 2 + -10");

        test_expression!(result, "1 < 2");
        test_expression!(result, "1 > 2");
        test_expression!(result, "1 < 1");
        test_expression!(result, "1 > 1");
        test_expression!(result, "1 == 1");
        test_expression!(result, "1 != 1");
        test_expression!(result, "1 == 2");
        test_expression!(result, "1 != 2");

        test_expression!(result, "if (true) { 10 }");
        test_expression!(result, "if (false) { 10 }");
        test_expression!(result, "if (1) { 10 }");
        test_expression!(result, "if (1 < 2) { 10 }");
        test_expression!(result, "if (1 > 2) { 10 }");
        test_expression!(result, "if (1 > 2) { 10 } else { 20 }");
        test_expression!(result, "if (1 < 2) { 10 } else { 20 }");

        test_expression!(result, "return 10;");
        test_expression!(result, "return 10; 9;");
        test_expression!(result, "return 2 * 5; 9;");
        test_expression!(result, "9; return 2 * 5; 9;");

        test_expression!(result, "if (10 > 1) { if (10 > 1) { return 10; } } return 1;");

        test_expression!(result, "5 + true;");
        test_expression!(result, "-true;");
        test_expression!(result, "true + false");
        test_expression!(result, "if (10 > 1) { true + false; }");

        test_expression!(result, "let x = 5");
        test_expression!(result, "let x = 3 > 5");
        test_expression!(result, "let x = 100 > 5");
        test_expression!(result, "let x = 100 > 5; x");
        test_expression!(result, "let x = 100 > 5; y");
        test_expression!(result, "let x = 100; let y = 200; x + y");

        insta::assert_snapshot!(result)
    }

    #[test]
    fn test_eval_function() {
        let mut result = String::new();

        test_expression!(result, "fn() { let x = 5 }");
        test_expression!(result, "fn(a, b, c) { let x = 5 }");
        test_expression!(result, "let a = 5; let c = fn() { a }; let b = 5; c;");
        test_expression!(result, "fn(a) { fn() { } }");
        test_expression!(result, "let a = 5; let c = fn() { a }; c();");
        test_expression!(result, "let a = 5; let c = fn(b) { b + 10 }; c(a);");
        test_expression!(result, "let a = 5; let c = fn(b) { b + 10 }; c(2);");
        test_expression!(result, "let o = 5; let a = fn() { o }; let b = fn(m) { m() }; b(a);");

        insta::assert_snapshot!(result)
    }

    #[test]
    fn test_eval_string() {
        let mut result = String::new();

        test_expression!(result, "\"This is a test\"");
        test_expression!(result, "\"This \" + \"is\\n\"");
        test_expression!(result, "\"This \" + 5");

        insta::assert_snapshot!(result)
    }

    #[test]
    fn test_len() {
        let mut result = String::new();

        test_expression!(result, "len(\"This is a test\")");
        test_expression!(result, "len(12345)");
        test_expression!(result, "len(\"This is\", \"a test\")");

        insta::assert_snapshot!(result)
    }

    #[test]
    fn test_array() {
        let mut result = String::new();

        test_expression!(result, "[0, 1, 3]");
        test_expression!(result, "[0, 1, 3][1]");
        test_expression!(result, "[0, 1, 3][-1]");
        test_expression!(result, "[0, 1, 3][5]");
        test_expression!(result, "let a = 2; [0, 1, 3][a]");
        test_expression!(result, "let myArray = [1, 2, 3]; myArray[0] + myArray[1] + myArray[2];");
        test_expression!(result, "let myArray = [1, 2, 3]; myArray[0] + myArray[myArray[0]] + myArray[myArray[0] + myArray[0]];");
        test_expression!(result, "len([1, 2, 3, 4])");
        test_expression!(result, "let myArray = [1, 2, 3]; len(myArray);");
        test_expression!(result, "first([1, 2, 3, 4])");
        test_expression!(result, "last([1, 2, 3, 4])");
        test_expression!(result, "push([1, 2], 5)");
        test_expression!(result, "rest([1, 2, 3])");

        insta::assert_snapshot!(result)
    }

    #[test]
    fn test_hash() {
        let mut result = String::new();

        test_expression!(result, "{\"a\": 5, \"b\": \"a\"}[\"b\"]");
        test_expression!(result, "{10: 5, 13: 5}[10]");
        test_expression!(result, "{true: 15, false: 65}[false]");
        test_expression!(result, "put({1: 15, 2: 65}, 3, 16)[3]");
        insta::assert_snapshot!(result)
    }
}