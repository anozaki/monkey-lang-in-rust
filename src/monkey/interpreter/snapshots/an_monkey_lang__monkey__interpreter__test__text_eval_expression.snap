---
source: src/monkey/interpreter/mod.rs
expression: result
---
Ok(Int(10))
Ok(Int(152))
Ok(Int(-5))
Ok(Int(-1523))
Ok(Bool(true))
Ok(Bool(false))
Ok(Bool(false))
Ok(Bool(true))
Ok(Bool(false))
Ok(Bool(true))
Ok(Bool(false))
Ok(Bool(true))
Ok(Bool(true))
Ok(Bool(true))
Ok(Bool(false))
Ok(Int(10))
Ok(Int(32))
Ok(Int(0))
Ok(Int(20))
Ok(Int(25))
Ok(Int(0))
Ok(Int(60))
Ok(Int(30))
Ok(Int(37))
Ok(Int(37))
Ok(Int(50))
Ok(Bool(true))
Ok(Bool(false))
Ok(Bool(false))
Ok(Bool(false))
Ok(Bool(true))
Ok(Bool(false))
Ok(Bool(false))
Ok(Bool(true))
Ok(Int(10))
Ok(Null)
Ok(Int(10))
Ok(Int(10))
Ok(Null)
Ok(Int(20))
Ok(Int(10))
Ok(Int(10))
Ok(Int(10))
Ok(Int(10))
Ok(Int(10))
Ok(Int(10))
Ok(Error("Expected number, got Int(5) and Bool(true)"))
Ok(Error("Unexpected value type: Bool(true)"))
Ok(Error("Expected number, got Bool(true) and Bool(false)"))
Ok(Error("Expected number, got Bool(true) and Bool(false)"))
Ok(Int(5))
Ok(Bool(false))
Ok(Bool(true))
Ok(Bool(true))
Ok(Error("identifier not found: Identifier(\"y\")"))
Ok(Int(300))

